### ISSUE TITLE
Weight is not even with leastconn

### ISSUE THREAD

**Issue by neobie** [haproxy/haproxy#1099]
![image](https://user-images.githubusercontent.com/4213973/106739930-bb43c980-6654-11eb-831a-f242e789aa19.png)

Ubuntu haproxy 2.0.13

I am having issue with leastconn connection. The session rate is not evenly distributed.
As you can see from the screenshots, both server is having same weight of 9.
But server 1 only receive half of the traffic/sessions compare to server 2. This cause my server load and cpu usage is unbalanced through the servers.

If I change the connection to "roundrobin", it is distributed evenly. 
I hope the **leastconn** issue can be fixed ?

**Comment by capflam** [comment]
First of all, the 2.0.13 is quite old. So please first upgrade to the 2.0.20 to be sure it is not an already fixed bug. If not, provide all information requested in the issue template. Thanks

**Comment by capflam** [comment]
Note also it could be good to report the CSV output instead of a screenshot. At least to have the different time values for your servers (connect/queue/response/total) and the connection reuse rate.

**Comment by TimWolla** [comment]
> First of all, the 2.0.13 is quite old. So please first upgrade to the 2.0.20 to be sure it is not an already fixed bug.

2.0.13 is the version shipped with Ubuntu Bionic.

@neobie You should use this PPA: https://haproxy.debian.net/#?distribution=Ubuntu&release=focal&version=2.0

**Comment by neobie** [comment]
Thanks for the guideline.
After updating to 2.0.20, the issue is still there.

From the ss, you can notice the server 3 and 4 is having same weight of 5, but the difference of sessions is so big.

![image](https://user-images.githubusercontent.com/4213973/106904846-a2abe000-6736-11eb-851a-7f5b157a6a10.png)

May I know what CSV can be provided ? What information you need to check on and where can I extract the information ?

**Comment by capflam** [comment]
In the CSV all the stats are visible. On a screenshot of the HTML stats page, many information are hidden in tooltips. Here, it could be good to have timing info. And you must also share your redacted configuration.

**Comment by wtarreau** [comment]
Wait a minute, here it's about `leastconn`, and it's expected. Leastconn doesn't balance the total traffic, it only makes sure no server is more loaded than other ones. And as you can see, your two servers with weight 5 had the exact same max number of sessions, and the two servers with weight 9 had the exact same as well. Further if you compare these two groups, you'll see that 16/9 == 9/5 which indeed shows that all your servers reached the exact same and balanced load.

What this shows here is that some of your servers are faster than the others, so they're taking more total requests. This is typically a case where users want to use leastconn. If you had used `roundrobin` instead, you'd have seen the exact same number of requests on your servers, but the slower ones would have taken a higher load.

Please have a look at the documentation for these keywords for more info.

**Comment by neobie** [comment]
@wtarreau Thank you for explanation.
Base on your explanation, leastconn only put a "max" cap so that the server not getting overwhelmed load. If the connection is low, they are equally balanced and not distributed according to weights defined.
If that is the case, which connection method can fulfil the roundrobin + leastconn (Distributing loads according to weights, while putting max cap according to weights) ?

**Comment by wtarreau** [comment]
No, leastconn doesn't put a "max", it always picks the least loaded. Servers are sorted by number of active connections (divided by their configured weight), that's how it works. In addition to this, leastconn does perform round-robin between servers having the same load. What you seem to want instead is to put a max on the servers load instead of using the weight. Then you should instead set a "maxconn" parameter on each server line, indicating how many max outstanding requests you want for each of them, and use roundrobin. Only servers that have not reached their limit will receive requests. Those having reached the limit will only pick pending requests once they finish processing one. If all servers are saturated, excess requests will be queued and picked by the first server to release a slot. This is heavily used as a way to prevent servers from using too many resources under highly varying loads.

**Comment by neobie** [comment]
Hi @wtarreau , I do want leastconn to perform a "roundrobin". But what I can see here is, **leastconn** make connections evenly distributed to all the servers, ignoring the "Weight" set. Please have a look at the table below. All the servers are equally distributed, no matter the weight set are **2, 4 or 6**. All **sessions** and **traffic in bytes** are equal.

![image](https://user-images.githubusercontent.com/4213973/107149395-527d8980-6993-11eb-9eed-b2d60ebc84e6.png)

**Comment by lukastribus** [comment]
As @wtarreau already explained above:

> Leastconn doesn't balance the total traffic, it only makes sure no server is more loaded than other ones. And as you can see, your two servers with weight 5 had the exact same max number of sessions, and the two servers with weight 9 had the exact same as well. Further if you compare these two groups, you'll see that 16/9 == 9/5 which indeed shows that all your servers reached the exact same and balanced load.

To verify whether or not leastconn is respecting your weight configuration, you need to look at Sessions->Max. And from that we can see that w-uc01-5, web01 and web03 all had 12 maximum sessions, while vu04 had 24 (double) and vu05 had 36 (triple) the amount of maximum sessions.

I therefor assume that the weight configuration matches the maximum sessions, with w-uc01-5, web01 and web03 having half the weight of vu04 and one third of the weight of vu05.


> All sessions and traffic in bytes are equal.

Total sessions and traffic is irrelevant for leastconn, as already explained.

**Comment by neobie** [comment]
@lukastribus , that's why I am asking, is there any rules to set a **roundrobin + **leastconn**** ? 
I need traffic to be distributed according to **roundrobin** weights while follow the **leastconn** weights.
That would be the ideal way to distribute the loads.
I need a **leastconn** to follows weight defined for **session max**  and also **session rate**. I hope this is very clear.

**Comment by wtarreau** [comment]
You seem to have a problem with what you exactly want. You're describing two opposite things. You're saying you want to have exactly the same number of requests per server, but exactly the same load. This makes no sense. **Either** you force servers to get the same number of requests **or** you force servers to get an equal load. But once you fix one, only the server's performance decides on the second one.

So let me restate it, hoping it becomes easier to understand this way:
  - roundrobin respects equal numbers of requests, regardless of the server's performance. This means that all servers will get the same number of requests and that slower servers will see higher concurrent requests since they take more time to complete them ;
  - leastconn respects equal numbers of concurrent requests. This means that no server will see more concurrent requests than another one, but that fastest servers will finish their requests faster and as such will process more requests
  - in addition to this, weights are **always** applied. For roundrobin, they allow you to adjust the ratios of requests. If you set weights 1,2,3,4 and send 10 requests, you'll respectively see 1,2,3,4 requests received on the servers. For leastconn they adjust the ratios of load. This means that server with weight 1 will be considered at the same load with 1 request as the server with weight 4 in presence of 4 requests.
  - you can still always limit the max number of concurrent connections or outstanding requests on a server using "maxconn", in which case, regardless of the algorithm in use, a server which reached the limit will not be selected by the algorithm.

In light of this, I hope ou finally understand that your sentence "*I need a leastconn to follows weight defined for session max and also session rate*" makes no sense. And it's not specific to the product but to the expectation itself: applied to a river, it would mean you want to be able to regulate **both** the water height and its speed; if it flows fast it's lower, if it flows slowly, it's higher.

**Comment by neobie** [comment]
@wtarreau 

_in addition to this, weights are always applied. For roundrobin, they allow you to adjust the ratios of requests. If you set weights 1,2,3,4 and send 10 requests, you'll respectively see 1,2,3,4 requests received on the servers. For leastconn they adjust the ratios of load. This means that server with weight 1 will be considered at the same load with 1 request as the server with weight 4 in presence of 4 requests._

The problem here is, leastconn **does not follow the **weight** set** . You can see the traffic and session rate from the screenshot I provided.

**Comment by wtarreau** [comment]
At this point I don't know what I can do if you refuse to read explanations.

Both Lukas and me showed you that it **does follow the weight in your captures**. Your capture shows 12 max concurrent requests for all servers with weight 2, 24 for all having weight 4, 36 for all having weight 6. Exactly as expected.

Please take the time to carefully read the explanations above. All you need is explained there.

**Comment by neobie** [comment]
![image](https://user-images.githubusercontent.com/4213973/108303954-ecd69c00-71e1-11eb-92b7-3a89a0be2f60.png)

![image](https://user-images.githubusercontent.com/4213973/108304064-26a7a280-71e2-11eb-8602-fb010f37eec7.png)


Just look at how unequal the traffic and loads are distributed, causing imbalance servers loads where web-uc02 above sits idle and relax though its weight is same as web-uc01 which busy serving requests and experience high loads.

**Comment by neobie** [comment]
@wtarreau you are only mentioning the "MAX" sessions,  but not the "AVERAGE" sessions distributed.
Please refer to the **Session rate (cur, max), Bytes in and out columns.**
I have no idea why the problem cant be seen. 
If you could, please explain why **web-uc02** received least traffic and server loads among 4 servers, and how to fix it so that it is receiving equal traffic, requests and distribution like **web-uc01**, as their weight is **14**.
Now even do05 and w-ln01 with weight of 11 is receiving more requests than web-uc01, leaving the web-uc01 servers server loads lowest among them.

**Comment by wtarreau** [comment]
Looking at your graphs clearly indicates that your servers are unequal, maybe they're sized differently, maybe they're also used by other applications, I don't know but as you can see, web-uc01 with **less** connections and a **lower** rate gets a *higher* CPU load, thus it's caused by the server itself and not the load balancer.
In addition what haproxy is doing is **exactly** what you asked it to do: send less **concurrent** connections to the last two than the first two, and make sure the first two have an **equal** number of concurrent connections. Your stats definitely show that this is well respected. Given that web-uc01 is significantly slower, this naturally results in **less total requests** to be processed for the same number of concurrent connections. And this is well translated in the session rate, byte in and out.

In fact, that's funny, because haproxy made you discover that you're having an obvious problem with this server that works much slower than other ones and you're still trying to figure why it gets less total traffic. In fact, leastconn here allowed your slower server to be used without degrading the delivered service. I don't know much what else to say, everything is working exactly as expected and documented. Maybe uc01 is in fact the same as the last two ones and ought to use a lower weight. Maybe it's running in a VM with noisy neighbors or an overloaded hypervisor, everything is possible, but now that you have all the elements to detect, prove and observe that there is a problem with this machine, I don't think anyone can be of more help to fix this server :-/

**Comment by neobie** [comment]
I am sorry to hear that you made such hurried conclusion.
Two servers web-uc01 and web-uc02 are of the same size, same instance, same CPU, same memory, from same providers.
So you jump into conclusions that these are servers issue not haproxy issues. 
I have totally out of idea what can I show you anymore.

**Comment by wtarreau** [comment]
Your stats **prove** the opposite. I can't say much more. Now you can decide to stop looking at stats if you want, but this is clear: for LOWER number of requests, one of them has HIGHER CPU usage. It's becoming difficult for me to give simpler explanations, I'm sorry, at this point I feel like I'm discussing politics or religion :-(

**Comment by neobie** [comment]
You cant even explain why web-uc02 is receiving less requests. When you are blaming it to have slower connections, why it is receiving 20+ connections only while all others are having 50+ connections ?

**Comment by wtarreau** [comment]
I explained it **MULTIPLE TIMES**. Can you read ? Or do you have a problem understanding basic concepts like waiting at the grocery ? You have a slower cashier and the rule you dictate is that you must have the same number of people waiting on all cashes. So at the end of the day, one of them processes less than the others. This is what you're having here **AND IT'S WRITTEN IN YOUR STATS OUTPUT**. If this is too difficult to understand, it means you don't want to use leastconn, which is what I already suggested 2 weeks ago. Just use roundrobin and let this faulty server further degrade your service. You'll be happy to see it process the same number of requests but then you'll complain that it gets a higher concurrent connection count (because it's slower).

**Comment by neobie** [comment]
You even suspect uc01 is using less weight before I made it to 14 and show the screenshot to you. What a sarcastic thought. Why would I do that ?

**Comment by neobie** [comment]
I think you are the one having difficulty of understanding, making all the hypothesis and conclusions yourself that doesn't exist.
So sorry to hear that.

**Comment by wtarreau** [comment]
At this point I prefer to give up. You provide yourself the proofs of all the explanations I give you and refuse to think and understand. I'm wondering if you're reading SMS, watching TV or driving your car while you're reading my explanations. Sincerely!

**Comment by wtarreau** [comment]
I suggest you find another product than haproxy. You'll be able to blame it for higher loads on web-uc01 (or 02 or whichever you want), and you'll fill their issue tracker, and they'll give you the same explanations but you'll annoy someone else.

**Comment by neobie** [comment]
![image](https://user-images.githubusercontent.com/4213973/108582827-921f7a80-7370-11eb-8f7c-54e2db9e2474.png)

Do explain if you can. Weight of 10 same loads with weight of 50. Same response time.

**Comment by TimWolla** [comment]
> Do explain if you can. Weight of 10 same loads with weight of 50. Same response time.

I'm trying one last time:

You have a total of 6 active sessions with more servers than sessions. No server has more than one active session. leastconn sends requests to the servers with the least connections (smallest number of sessions). I'd say it is working fine.

------------------

If you _still_ insist that there's anything wrong there, then:

File a new issue and fill out the issue template in full. This includes the full unredacted configuration, the output of `haproxy -vv` and a full specification of the environment (e.g. hardware, interconnection, workload, etc.).

Guessing what you want from a screenshot with nothing except "this is not working" is not a useful way to spend the time of the developers.

I'm locking this issue at this point. If any further exchange is desired from your side then I expect an issue that includes all of the necessary information.
